cupy/core/fusion.pyx

```diff
@@ -798,7 +801,7 @@ class _FusionHistory(object):
         else:
             _, reduce_expr, postmap_expr, reduce_ctype = self.reduce_op.routine
             if reduce_ctype is None:
-                reduce_ctype = 'type_in0_raw'
+                reduce_ctype = 'type_out0_raw'
```

Comment:
This does not make any semantics difference if CUB is not in use (note that in fusion `type_in0_raw` and `type_out0_raw` are always identical, for a reason I don't understand yet), but if CUB is turned on then we must follow `ReductionKernel`'s expectation that if the reduction type is not specified, we use the output dtype.

@asi1024 It seems you were the last one touching this part of code. Any chance you could recall why `type_in0_raw ` was used here?

Reply:
Sorry, that is still under investigation, but I'm now wondering that `reduce_ctype` should be neither `type_in0_raw` nor `type_out0_raw`. `reduce_ctype` is the type of data after pre-map and before post-map. It may be a different type from both input type and output type.

Reply:
That's right. If `reduce_ctype` is given, we need to use it (ex: `max`/`min`); if not given, we use the output dtype (ex: `sum`/`prod`). My change would make fusion in line with everywhere in the codebase.

Reply:
In the fusion case, though, I think it's expected that the type before and after postmap is identical, another part I don't understand but verify via tests.